Cookbook/Proposals/ManagingAssets.md

```diff
@@ -0,0 +1,58 @@
+1. All new icons and images we are going to use should be added to `BabylonDependencies.framework` in `Assets.xcassets` catalog 
+2. `struct DesignLibrary.Tokens` should have another property for icons `public let icons: Icons`
+3. Then we have to add a new case to `enum ImageIdentifier` which will be embedded in `struct Icons`.
+4. To access e.g. close icon we could use subscripts `designLibrary.tokens.icons[.close]`
+5. If we are updating an icon we should check if any other place uses it. If so and that other place shouldn't be updated we should create a new icon with an updated design.
+6. If we are removing some code which is using some icons we should check if that code was the last place which was using given icon, if so icon should be deleted from `Assets.xcassets`.
+7. To support the overriding of standard icons in white label apps we should update `ImageCatalogueAware.image(for imageIdentifier:in bundle: compatibleWith traitCollection:) -> UIImage` to firstly access image from `Bundle.main` and then fallback to `BabylonDependencies` if image was not overridden. We can also add support to specify from which location we would like to access given image.
+
+## Impact on existing codebase
+
+Unfortunately, in our codebase icons and images are located in different places. If we will agree on above set of rules, they should be applied for newly created assets. We should also make an effort to eliminate technical debt and migrate all existing icons into `BabylonDependencies`.
+Keeping every icon and image in one location potentially shouldn’t increase the size of the target application on the condition that the app has all our frameworks linked.
+
+## Alternatives considered
+
+1. We could try to systematize the way we include assets in specific feature frameworks but it can cause problems described in the motivation section.
+
+2. Instead of accessing icons or images by subscripts `designLibrary.tokens.icons.iconography[.close]` we could use new feature of Swift 5.1 `@dynamicMemberLookup` which could be combined with `KeyPath`. Then we could write just `designLibrary.tokens.icons.close`.
+To achieve that firstly we have to mark `struct Icons` with `@dynamicMemberLookup`. Then `ImageIdentifier` has to become `struct` with `String` properties with default value:
+```
+struct ImageIdentifier {
+    let close = "close"
+}
+```
+Later we need create property `let imageIdentifier = ImageIdentifier()` inside `struct Icons`. Having all of these we can finally write dynamicMember subscript:
+```
+subscript(dynamicMember keyPath: KeyPath<ImageIdentifier, String>) -> UIImage {
+    return Icons.image(for: imageIdentifier[keyPath: keyPath])
+}
+```
+And finally use it like: `designLibrary.tokens.icons.close`. The only drawback of this approach is the fact that for new icons we have to create new property and assign the default value to it, compared to just create new `enum` `case` with the name matching asset name.
```

Comment:
You describe the drawback here, but what are the benefits of doing that, especially compared to what you proposed in the previous section (line 28) that would directly allow `designLibrary.tokens.icons.close`?

Reply:
The benefit here a little better syntax (keeping addition simple) but it's a matter of taste.

Reply:
But you already have the smaller syntax on line 28 though

Reply:
`designLibrary.tokens.icons[.close]`
`designLibrary.tokens.icons.close`
Second one is shorter and the second one uses `@dynamicMemberLookup`. 
If I misunderstand you we should find the same perspective here.

Reply:
Ohhh those square brackets were supposed to be real, literal, square brackets, of course 🤦‍♂️

Idk why but I read that `designlibrary.tokens.icons[.close]` on line 28 as some syntax supposed to mean "`designlibrary.tokens.icons.close` but replace `.close` here with whatever icon name you want", as if the square brackets were not part of the actual Swift expression but were there to explain the presence of a placeholder… so that's why I didn't read it as if it were literal square brackets here.

But in fact your suggestion is to implement `subscript` on `Icon` there, makes sense now, don't know how I didn't see that on first read, my bad (maybe mentioning `subscript` on the proposed solution around line 28 could help?)

Then maybe we could add my suggestion / what I thought it would be in the first place (make `struct Icon` define its icons using `static let close = …`) in the alternatives considered?

Reply:
Sure I will add it :)

Reply:
I would rather add alternative as computed static var `static var close: UIImage`. `static let` is lazy loaded and after that kept in memory. I don't think we would like to flood memory with many `UIImage`s

Reply:
Hi @AliSoftware I extended SwiftGen section. Please let me know if you still would like me to add `struct Icon` and  `static let close = ...` I'm not sure if it could be a simpler approach then what is proposed is SwiftGen section but you can try to convince me :)